process/sync/shardForkDetector.go

```diff
@@ -0,0 +1,174 @@
+package sync
+
+import (
+	"bytes"
+	"math"
+	"strings"
+
+	"github.com/ElrondNetwork/elrond-go/consensus"
+	"github.com/ElrondNetwork/elrond-go/data"
+	"github.com/ElrondNetwork/elrond-go/process"
+)
+
+// shardForkDetector implements the shard fork detector mechanism
+type shardForkDetector struct {
+	*baseForkDetector
+}
+
+// NewShardForkDetector method creates a new shardForkDetector object
+func NewShardForkDetector(rounder consensus.Rounder) (*shardForkDetector, error) {
+	if rounder == nil || rounder.IsInterfaceNil() {
+		return nil, process.ErrNilRounder
+	}
+
+	bfd := &baseForkDetector{
+		rounder: rounder,
+	}
+
+	bfd.headers = make(map[uint64][]*headerInfo)
+	checkpoint := &checkpointInfo{}
+	bfd.setFinalCheckpoint(checkpoint)
+	bfd.addCheckpoint(checkpoint)
+
+	sfd := shardForkDetector{
+		baseForkDetector: bfd,
+	}
+
+	return &sfd, nil
+}
+
+// AddHeader method adds a new header to headers map
+func (sfd *shardForkDetector) AddHeader(
+	header data.HeaderHandler,
+	headerHash []byte,
+	state process.BlockHeaderState,
+	finalHeaders []data.HeaderHandler,
+	finalHeadersHashes [][]byte,
+) error {
+
+	if header == nil || header.IsInterfaceNil() {
+		return ErrNilHeader
+	}
+	if headerHash == nil {
+		return ErrNilHash
+	}
+
+	err := sfd.checkBlockBasicValidity(header, state)
+	if err != nil {
+		return err
+	}
+
+	if state == process.BHProcessed {
+		// add final header(s), add check point and remove all the past checkpoints/headers
+		sfd.addFinalHeaders(finalHeaders, finalHeadersHashes)
+		sfd.addCheckpoint(&checkpointInfo{nonce: header.GetNonce(), round: header.GetRound()})
+		sfd.removePastOrInvalidRecords()
+	}
+
+	sfd.append(&headerInfo{
+		nonce: header.GetNonce(),
+		round: header.GetRound(),
+		hash:  headerHash,
+		state: state,
+	})
+
+	probableHighestNonce := sfd.computeProbableHighestNonce()
+	sfd.setLastBlockRound(uint64(sfd.rounder.Index()))
+	sfd.setProbableHighestNonce(probableHighestNonce)
+
+	return nil
+}
+
+func (sfd *shardForkDetector) addFinalHeaders(finalHeaders []data.HeaderHandler, finalHeadersHashes [][]byte) {
+	finalCheckpointWasSet := false
+	for i := 0; i < len(finalHeaders); i++ {
+		isFinalHeaderNonceHigherThanCurrent := finalHeaders[i].GetNonce() > sfd.GetHighestFinalBlockNonce()
+		if isFinalHeaderNonceHigherThanCurrent {
+			if !finalCheckpointWasSet {
+				sfd.setFinalCheckpoint(&checkpointInfo{nonce: finalHeaders[i].GetNonce(), round: finalHeaders[i].GetRound()})
+				finalCheckpointWasSet = true
+			}
+
+			sfd.append(&headerInfo{
+				nonce: finalHeaders[i].GetNonce(),
+				round: finalHeaders[i].GetRound(),
+				hash:  finalHeadersHashes[i],
+				state: process.BHNotarized,
+			})
+		}
+	}
+}
+
+// CheckFork method checks if the node could be on the fork
+func (sfd *shardForkDetector) CheckFork() (bool, uint64, []byte) {
```

Comment:
I think there is still an edge case that is not covered.
For example what happens when there are two competing shard blocks on a nonce and metachain did not notarize any of these blocks? 
since you have no rule to choose between the two which is consistent across nodes in the shard then you might end up with a partitioned shard, and with consensus that cannot produce a block for unspecified number of rounds. 
If we want to introduce also possibility of ending an epoch early when one shard is not producing blocks for a certain number of rounds (this is in discussion right now), this could introduce huge penalties.

The simple fork choice rules for shard as I see them would be:
1. If metachain has notarized a block for that nonce then that is the choice and the other blocks can be dismissed/removed 
2. If there is no block notarized by the metachain then go with local rule, but don't dismiss the other blocks, as you might need to go back to 1.

Reply:
Yes, good catch here. Actually the flow of this algorithm was exactly as you described it above (so I have put it back and now we have again only one method CheckFork for both of them shard/meta). The single difference was that on roll back all the blocks which would be rolled back would have been removed from storage. Now they will not be deleted anymore, and so we can manage the situation in which metachain would have notarized something on which we didn't construct and we should roll back and go on that way.